Editorial review: Add moveBefore() docs (#39002)
mdn/content: files/en-us/web/api/documentfragment/movebefore/index.md

---
title: "DocumentFragment: moveBefore() method"
short-title: moveBefore()
slug: Web/API/DocumentFragment/moveBefore
page-type: web-api-instance-method
browser-compat: api.DocumentFragment.moveBefore
---

{{APIRef("DOM")}}

The **`moveBefore()`** method of the {{domxref("DocumentFragment")}} interface moves a given {{domxref("Node")}} inside the invoking `DocumentFragment` as a direct child, before a given reference node.

## Syntax

```js-nolint
moveBefore(movedNode, referenceNode)
```

### Parameters

- `movedNode`

>   - : A {{domxref("Node")}} representing the node to be moved.

Reviewer: ```suggestion
  - : A {{domxref("Node")}} representing the node to be moved. Note that this must be an {{domxref("Element")}} or a {{domxref("CharacterData")}} node.
```

---
title: "DocumentFragment: moveBefore() method"
short-title: moveBefore()
slug: Web/API/DocumentFragment/moveBefore
page-type: web-api-instance-method
browser-compat: api.DocumentFragment.moveBefore
---

{{APIRef("DOM")}}

The **`moveBefore()`** method of the {{domxref("DocumentFragment")}} interface moves a given {{domxref("Node")}} inside the invoking `DocumentFragment` as a direct child, before a given reference node.

## Syntax

```js-nolint
moveBefore(movedNode, referenceNode)
```

### Parameters

- `movedNode`
  - : A {{domxref("Node")}} representing the node to be moved.
- `referenceNode`
  - : A {{domxref("Node")}} that `movedNode` will be moved before, or `null`. If the value is `null`, `movedNode` is inserted at the end of the invoking `DocumentFragment`'s child nodes.
    > [!NOTE]
    > The `referenceNode` parameter is mandatory. You must explicitly pass a {{domxref("Node")}} or `null`.

### Return value

> None ({{jsxref("Undefined")}}).

Author: Updated on all three pages.

### Exceptions

- `HierarchyRequestError` {{jsxref("TypeError")}}
  - : Thrown when:
    - The specified `referenceNode` is not yet added to the DOM, and you are trying to move it inside a node that is part of the DOM, or vice versa.

>     - An attempt is made to move `movedNode` between two different documents.

Author: Ah, good point. These need updating. I did some testing, and edited them to:

- The specified `referenceNode` is already added to the DOM, and you are trying to move it inside a `DocumentFragment`.
- You are trying to move `movedNode` between two different document fragments.

I found that both of these cases produced the exception in question.

- `NotFoundError` {{jsxref("TypeError")}}
  - : The specified `referenceNode` is not a child of the `DocumentFragment` you are calling `moveBefore()` on, that is, the fragment you are trying to move `movedNode` inside.
- `TypeError` {{jsxref("TypeError")}}
  - : The second argument was not supplied.

## Description

> The `moveBefore()` method moves a given node to a new place in the DOM. It provides similar functionality to the {{domxref("Node.insertBefore()")}} method, except that it doesn't remove and then reinsert the node. This means that the state of the node (which would be reset if moving it with `insertBefore()` and similar mechanisms) is preserved after the move. This includes:

Author: Oops, updated to `DocumentFragment`

- [Animation](/en-US/docs/Web/CSS/CSS_animations) and [transition](/en-US/docs/Web/CSS/CSS_transitions) state.
- {{htmlelement("iframe")}} loading state.
- Interactivity states (for example, {{cssxref(":focus")}} and {{cssxref(":active")}}).
- [Fullscreen](/en-US/docs/Web/API/Fullscreen_API) element state.
- Open/close state of [popovers](/en-US/docs/Web/API/Popover_API).
- Modal state of {{htmlelement("dialog")}} elements (modal dialog will not be closed).

The play state of {{htmlelement("video")}} and {{htmlelement("audio")}} elements is not included in the above list, as these elements retain their state when removed and reinserted, regardless of the mechanism used.

> [!NOTE]
> When observing changes to the DOM using a {{domxref("MutationObserver")}}, nodes moved with `moveBefore()` will be recorded with a [removed node](/en-US/docs/Web/API/MutationRecord/removedNodes) and an [added node](/en-US/docs/Web/API/MutationRecord/addedNodes).

### `moveBefore()` constraints

There are some constraints to be aware of when using `moveBefore()`:

> - It can only work when moving a node within the same document.

Author: updated to document fragment

- It won't work if you try to move a node that is not connected to the DOM to an already connected parent, or vice versa.

In such cases, `moveBefore()` will fail with a `HierarchyRequestError` exception. If the above constraints are requirements for your particular use case, you should use {{domxref("Node.insertBefore()")}} instead, or use [`try...catch`](/en-US/docs/Web/JavaScript/Reference/Statements/try...catch) to handle the errors that arise from such cases.

> ### Moving custom elements

Reviewer: FWIW I don't really understand why `connectedCallback()` and `disconnectedCallback()` are fired when using `moveBefore()`.

> Each time a [custom element](/en-US/docs/Web/API/Web_components/Using_custom_elements) is moved via `moveBefore()` (or similar methods such as {{domxref("Node.insertBefore()")}}), the `disconnectedCallback()` and `connectedCallback()` [lifecycle callbacks](/en-US/docs/Web/API/Web_components/Using_custom_elements#custom_element_lifecycle_callbacks) are fired.
> 
> Consider this minimal example:
> 
> ```js
> class MyCustomElement extends HTMLElement {
>   constructor() {
>     super();
>   }
> 
>   connectedCallback() {
>     console.log("Custom element added to page.");
>   }
> 
>   disconnectedCallback() {
>     console.log("Custom element removed from page.");
>   }
> }
> 
> customElements.define("my-custom-element", MyCustomElement);
> ```
> 
> In this case, "Custom element removed from page." and "Custom element added to page." are logged to the console with each move. This might be your intended behavior. However, if you use the callbacks to run initialization and cleanup code, it might cause problems with the state of the moved element.
> 
> Custom elements can be opted in to state-preserving moves via the `connectedMoveCallback()` lifecycle callback. If defined in the constructor, this will run instead of `connectedCallback()` and `disconnectedCallback()` when an element instance is moved via `moveBefore()`. You could add an empty `connectedMoveCallback()` to stop the other two callbacks running, or include some custom logic to handle the move:
> 
> ```js
> connectedMoveCallback() {
>   console.log("Custom element moved with moveBefore()");
> }
> ```

Reviewer: I think this could be more concise and sort of buries the lede, especially with:

> This might be your intended behavior. However, if you use the callbacks to run initialization and cleanup code, it might cause problems with the state of the moved element. 

...when AIUI these callbacks are primarily intended for setup and teardown code. From the guide:

> The specification recommends that, as far as possible, developers should implement custom element setup in this callback

So I think it's worth considering sharpening the language here, something like:

```suggestion
Each time a [custom element](/en-US/docs/Web/API/Web_components/Using_custom_elements) is moved via methods such as {{domxref("Node.insertBefore()")}}), the `disconnectedCallback()` and `connectedCallback()` [lifecycle callbacks](/en-US/docs/Web/API/Web_components/Using_custom_elements#custom_element_lifecycle_callbacks) are fired, because the element is being removed from and then added back to the DOM.

These callbacks are typically intended to allow you to implement any initialization or cleanup code.

The same callbacks are, by default, fired when the element is moved using `moveBefore()`. However, since the reason to use `moveBefore()` is that the element's state is preserved, it may not be appropriate to run initialization and setup code. To handle this situation, you can define the `connectedMoveCallback()` for the element: if you do, then neither `disconnectedCallback()` or `connectedCallback()` are called when the element is moved using `moveBefore()`.
```

Author: I've used some of your wording.

Reviewer: I still think this https://pr39002.review.mdn.allizom.net/en-US/docs/Web/API/Web_components/Using_custom_elements#moving_custom_elements_lifecycle_side-effects is confusing.

AIUI, this issue, and the solution of implementing `connectedMoveCallback()`, is specific to `moveBefore()`. It's not considered an issue for `insertNode()`, because in that case the element actually is removed and reinserted, and its state is reset. And, in that situation,  implementing `connectedMoveCallback()` does not help you. Is that right?

But this section doesn't say that. It says you can move elements using either API, and that the `disconnectedCallback`/`connectedCallback` will fire in these situations. It then says:

> since these callbacks are typically used to implement any required initialization or cleanup code, moving the element may cause problems with its state.

It doesn't indicate that this is specifically an issue for `moveBefore()`. And "may cause problems with its state" is vague. What we mean AIUI is: if you move an element using `moveBefore()`, its state is preserved, so you don't want to run cleanup/initialization code. This is IMO a much more precise way to express it.

It then says:

> Custom elements can be opted in to state-preserving moves via the [Element.moveBefore()](https://pr39002.review.mdn.allizom.net/en-US/docs/Web/API/Element/moveBefore) method and the connectedMoveCallback() lifecycle callback.

...but it's not at all obvious what "opted in to state-preserving moves" means. For example, calling `moveBefore()` doesn't "opt into" a state preserving move, it performs a state preserving move. It would be much clearer to say something like "In contrast to `insertNode()`, if you move an element using `moveBefore()` then its state is preserved".

FWIW I also don't see the value in the example, which is just an entirely generic presentation of the callbacks, as a cut down copy/paste of the example in https://pr39002.review.mdn.allizom.net/en-US/docs/Web/API/Web_components/Using_custom_elements#custom_element_lifecycle_callbacks. And I think it's odd in that context to say "This might be your intended behavior." - surely logging `"Custom element added to page."` is almost never going to be the intended behavior.

Anyway, these are the kinds of considerations I had in mind when making my comment. I don't want this to be like "you must make exactly the change I've suggested" - I'm happy to discuss. But I'd like to understand why you're not keen on accepting the changes I've suggested.

Author: I didn't use all of your wording because the last paragraph seemed a bit wordy and bottom heavy.

I've taken all of your follow-up comments here on board, and rewritten the section quite heavily. I do think the state stuff is an issue for `insertBefore()` (and other similar DOM move methods), because they are how people have moved elements for a long time before `moveBefore()` was created.

I want to be careful to frame it as an issue with `moveBefore()` and other similar move methods, which can be fixed using the `connectedMoveCallback()` lifecycle callback, specifically when you use `moveBefore()` method to move elements. 

Rather than just an issue with `moveBefore()`.

> AIUI, this issue, and the solution of implementing connectedMoveCallback(), is specific to moveBefore(). It's not considered an issue for insertNode(), because in that case the element actually is removed and reinserted, and its state is reset. And, with insertNode(), connectedMoveCallback() does not suppress the other callbacks. Is that right?

Your summary is correct, but it _is_ still an issue for people using `insertBefore()` when they are trying to use it to move nodes rather than insert them, which is often the case.

Reviewer: I think this version is great. 👍

## Examples

### Basic `moveBefore()` usage

In this demo we illustrate basic usage of `moveBefore()`.

#### HTML

> The HTML features three {{htmlelement("button")}} elements, and an {{htmlelement("article")}} element. We will use the buttons to control inserting `DocumentFragments` into the `<article>`, and emptying it.

Reviewer: ```suggestion
The HTML features three {{htmlelement("button")}} elements, and an {{htmlelement("article")}} element. We will use the buttons to control inserting `DocumentFragments` into the `<article>` and emptying it.
```

Also though, I find it grating to mix identifiers into prose like "`DocumentFragment`s", rather than something like "`DocumentFragment` instances". This is a very pedantic comment but I have checked and at least Hamish feels the same way.

Author: Well, in this case, it introduced an actual mistake. Updated to "...inserting `DocumentFragment` instances...".

And I also removed the comma as per the suggestion.

<article id="wrapper"></article>
```

#### CSS

We provide some rudimentary styling for the look and feel and spacing of elements that will later be inserted into the page as children of JavaScript-generated `DocumentFragment`s.

```css live-sample___movebefore-basic
#section1,
#section2,
#mover {
  display: inline-block;
  width: 200px;
  height: 30px;
  border: 5px solid rgb(0 0 0 / 0.25);
  margin-top: 10px;
}

#section1,
#section2 {
  background-color: hotpink;
}

#mover {
  background-color: orange;
}
```

#### JavaScript

In our script, we define a function, `createFragment()`, which creates a `DocumentFragment` containing a {{htmlelement("div")}} element and two {{htmlelement("section")}} elements as immediate children.

We then attach a click event listener to each `<button>` via {{domxref("EventTarget.addEventListener", "addEventListener()")}}:

- The first button appends the `DocumentFragment` to the `#wrapper` `<article>` element, unmodified.
- The second button appends the `DocumentFragment` to the `#wrapper` `<article>` element, but first uses `moveBefore()` to move the `<div>` to be the second child of the `DocumentFragment` rather than the first.

> - The third button empties the `#wrapper` `<article>` element using {{domxref("Element.innerHTML", "innerHTML()")}}.

Reviewer: ```suggestion
- The third button empties the `#wrapper` `<article>` element using {{domxref("Element.innerHTML", "innerHTML")}}.
```

```js live-sample___movebefore-basic
const wrapper = document.getElementById("wrapper");
const insertBtn1 = document.getElementById("insert1");
const insertBtn2 = document.getElementById("insert2");
const clearBtn = document.getElementById("clear");

function createFragment() {

>   let fragment = new DocumentFragment();

Reviewer: ```suggestion
  const fragment = new DocumentFragment();
```

- [Animation](/en-US/docs/Web/CSS/CSS_animations) and [transition](/en-US/docs/Web/CSS/CSS_transitions) state.
- {{htmlelement("iframe")}} loading state.
- Interactive node {{cssxref(":focus")}} and {{cssxref(":active")}} states.
- [Fullscreen](/en-US/docs/Web/API/Fullscreen_API) element state.
- Open/close state of [popovers](/en-US/docs/Web/API/Popover_API).
- Modal state of {{htmlelement("dialog")}} elements (modal dialog will not be closed).

The play state of {{htmlelement("video")}} and {{htmlelement("audio")}} elements is not included in the above list, as these elements retain their state when removed and reinserted, regardless of the mechanism used.

> [!NOTE]
> When observing changes to the DOM using a {{domxref("MutationObserver")}}, nodes moved with `moveBefore()` will be recorded with a [removed node](/en-US/docs/Web/API/MutationRecord/removedNodes) and an [added node](/en-US/docs/Web/API/MutationRecord/addedNodes).

> ### `moveBefore()` constaints

Reviewer 2: constraints

Author: Oops! Fixed.

There are some constraints to be aware of when using `moveBefore()`:

- It can only work when moving a node within the same document.
- It won't work if you try to move a node that is not connected to the DOM to an already connected parent, or vice versa.

In such cases, `moveBefore()` will fail with a `HierarchyRequestError` exception. If the above constraints are requirements for your particular use case, you should use {{domxref("Node.insertBefore()")}} instead, or use [`try...catch`](/en-US/docs/Web/JavaScript/Reference/Statements/try...catch) to handle the errors that arise from such cases.

### Moving custom elements

> When moving a [custom element](/en-US/docs/Web/API/Web_components/Using_custom_elements), care needs to be taken to handle the `connectedCallback()` and `disconnectedCallback()` [lifecycle callbacks](/en-US/docs/Web/API/Web_components/Using_custom_elements#custom_element_lifecycle_callbacks) appropriately, if any are defined in the custom element's constructor.

Reviewer 2: "care needs to be taken" is not the right framing here.
By default, custom elements don't support state-preserving move, and their `disconnected` and `connected` callbacks are fired.

Authors of web components can *opt in* to state-preserving move behavior, by implementing the `connectedMoveCallback`, which would be called instead of the disconnected+connected default.

Note: this default behavior is introduced in order to maintain backwards compatibility with existing custom elements who might rely on being disconnected and reconnected when moved around the DOM.

Author: OK, I get this now; thanks for the explanation. I've reworded the section as follows:

---

### Moving custom elements

Each time a [custom element](/en-US/docs/Web/API/Web_components/Using_custom_elements) is moved via `moveBefore()` (or similar methods such as {{domxref("Node.insertBefore()")}}), the `disconnectedCallback()` and `connectedCallback()` [lifecycle callbacks](/en-US/docs/Web/API/Web_components/Using_custom_elements#custom_element_lifecycle_callbacks) are fired.

Consider this minimal example:

```js
class MyCustomElement extends HTMLElement {
  constructor() {
    super();
  }

  connectedCallback() {
    console.log("Custom element added to page.");
  }

  disconnectedCallback() {
    console.log("Custom element removed from page.");
  }
}

customElements.define("my-custom-element", MyCustomElement);
```

In this case, "Custom element removed from page." and "Custom element added to page." are logged to the console with each move. This might be your intended behavior. However, if you use the callbacks to run initialization and cleanup code, it might cause problems with the state of the moved element.

Custom elements can be opted in to state-preserving moves via the `connectedMoveCallback()` lifecycle callback. If defined in the constructor, this will run instead of `connectedCallback()` and `disconnectedCallback()` when an element instance is moved via `moveBefore()`. You could add an empty `connectedMoveCallback()` to stop the other two callbacks running, or include some custom logic to handle the move:

```js
connectedMoveCallback() {
  console.log("Custom element moved with moveBefore()");
}
```

Reviewer 2: Perfect thanks!
